builtin_array.go

```diff
@@ -786,6 +786,7 @@ func (r *Runtime) arrayproto_map(call FunctionCall) Value {
 	if _, stdSrc := o.self.(*arrayObject); stdSrc {
 		if arr, ok := a.self.(*arrayObject); ok {
 			values := make([]Value, length)
+			r.vm.tmpValues = values
```

Comment:
We assign a reference to the values array so that we are now tracking memory utilization of this temporary `values` array.

Reply:
Is there any chance this might be used concurrently (I don't think so but you might know better at this point)? I haven't looked into it but just thought about it as part of the changes we made recently in our http2 package and wanted to be careful since this is now shared from the runtime.

Reply:
Yeah so this is one of the main reasons that I steered away from a previous approach that consisted of putting the temporary objects directly onto the stack. I suspect this is possible but not something we'd commonly encounter.

I think an operation like the one below might give us concurrent usage like you're suggesting
```
Promise.all([arr1.map(asyncFunc), arr2.map(asyncFunc)])
```

The end result is that we'd only be able to track a single map operation at a time. Unless we do something like create a unique hash for the map operation and track the values array under that specific hash.

Reply:
JS should be single-threaded, so this function should never be used concurrently within the same runtime unless we have another bug. I think it's safe to assume that this code will run on a single thread. Other members would run into data races as well if that wasn't the case.

Reply:
Yeah we're ok here

```
exports = async function(arg){
  var arr = [1,2,3,4,5]

  await Promise.all([
    arr.map(async x => { console.log(x) }),
    arr.map(async x => { console.log(x) }),
    arr.map(async x => { console.log(x) }),
  ])
};
```

```
> logs: 
1
2
3
4
5
1
2
3
4
5
1
2
3
4
5
```

Reply:
My concern was that even though it is single threaded, that it would not finish the entire map operation and bounce around.

Reply:
Wait.. I found a way to recreate what I was worried about

```
exports = async function(arg){
  var arr = [1,2,3,4,5]

  await Promise.all([
    arr.map(async x => { setTimeout(() => { console.log(x) }, 1000 - x) }),
    arr.map(async x => { setTimeout(() => { console.log(x) }, 1000 - x) }),
    arr.map(async x => { setTimeout(() => { console.log(x) }, 1000 - x) }),
  ])
};
```

```
> took 1.272834522s
> logs: 
5
5
5
4
3
4
4
2
2
3
3
2
1
1
1
```

Reply:
![image](https://github.com/mongodb-forks/goja/assets/6693710/223111eb-05de-4a3e-981a-593df022f30a)

Reply:
I'm going to create a follow-up to address this concern so that we can go ahead and address the immediate concern of map memory usage because we know we've OOMed because of this.

cc: @arahmanan for visibility because I know we briefly discussed potential solutions.

Reply:
https://jira.mongodb.org/browse/BAAS-32408